File: NeetCode/Binary Search/BinarySearch.java
public class BinarySearch {
    public static int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length-1;

        while(left <= right) {                              //! Time Complexity = O(log n)
            int middle = left + (right - left)/2;
            if(target == nums[middle]) {
                return middle;
            } else if(target < nums[middle]) {
                right = middle-1;
            } else {
                left = middle+1;
            }
        }
        
        return -1;
    }                                                       //! Total Time Complexity = O(log n)

    public static void main(String[] args) {
        // int[] nums = {-1, 0, 3, 5, 9, 12};
        // int target = 9;

        int[] nums = {-1, 0, 3, 5, 9, 12};
        int target = 2;

        System.out.println(search(nums, target));
    }    
}

File: NeetCode/Binary Search/BinarySearch.cs
public class Solution {
    public int Search(int[] nums, int target) {
        return BinarySearch(0, nums.Length-1, nums, target);
    }

    private int BinarySearch(int left, int right, int[] nums, int target) {
        if (left > right)
            return -1;

        int mid = left + (right - left) / 2;

        if (nums[mid] == target) {
            return mid;
        } else if (nums[mid] < target) {
            return BinarySearch(mid + 1, right, nums, target);
        } else {
            return BinarySearch(left, mid - 1, nums, target);
        } 
    }
}
